### ISSUE TITLE
Materialized view for postgres replication source crashed on delete

### ISSUE THREAD

**Issue by SephVelut** [MaterializeInc/materialize#12783]
### What version of Materialize are you using?

0.26.2

### How did you install Materialize?

Docker image

### What is the issue?

I have one materialized view created for a postgres source. I'll explain the query. The query's inner select will order rows by `xxx_id` and `created_at` in desc. Then it selects on the first distinct `xxx_id`, in effect, deduping the duplicate `xxx_id` and picking the one with the latest `created_at`. The outer query just bins the rows by hourly segments and does a group by to obtain an hourly sum.

```sql
create materialized view total_payout as
select
	vertical,
  	sum(renumeration),
  	date_bin('1 hour', to_timestamp(xxx_timestamp), '1970-01-01') as xxx_timestamp_bin
from
(select distinct on (xxx_id)
  xxx_id,
  xxx_timestamp,
  vertical,
  renumeration
from
	xxx_receipts
where 
	vertical in ('XXX', 'XXX XXXX')
	and renumeration > 0
order by 
	xxx_id, created_at desc
) q
group by
	vertical,
	xxx_timestamp_bin
```

This works fine. Materialize replicates, I can query the view just fine. Up until I delete a record in the source table in the database materialize is replicating from. Then materialize crashes after about 3 seconds. Log is below.

### Relevant log output

```source
2022-05-31T16:21:52.914614Z ERROR panic: assertion failed: diff > 0

thread: timely:work-0

location: src/dataflow/src/compute/render/top_k.rs:287:21

version: 0.26.2 (10ff4ae61d835809fbc758450d519cd4b48929c4)

backtrace:

   0: materialized::handle_panic

             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0335afc44bffe5852-1/materialize/tests/src/materialized/src/bin/materialized/main.rs:1005:21

   1: core::ops::function::Fn::call

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/ops/function.rs:70:5

   2: std::panicking::rust_panic_with_hook

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:610:17

   3: std::panicking::begin_panic_handler::{{closure}}

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:500:13

   4: std::sys_common::backtrace::__rust_end_short_backtrace

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys_common/backtrace.rs:139:18

   5: rust_begin_unwind

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:498:5

   6: core::panicking::panic_fmt

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:116:14

   7: core::panicking::panic

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:48:5

   8: mz_dataflow::compute::render::top_k::<impl mz_dataflow::compute::render::context::Context<G,mz_repr::row::Row>>::render_topk::render_top1_monotonic::{{closure}}

             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0335afc44bffe5852-1/materialize/tests/src/dataflow/src/compute/render/top_k.rs:287:21

      <timely::dataflow::stream::StreamCore<S,alloc::vec::Vec<D>> as timely::dataflow::operators::map::Map<S,D>>::map::{{closure}}::{{closure}}::{{closure}}::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/map.rs:57:78

      core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/ops/function.rs:280:13

      core::option::Option<T>::map

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/option.rs:883:29

      <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/iter/adapters/map.rs:103:9

      timely::dataflow::channels::pushers::buffer::Session<T,alloc::vec::Vec<D>,P>::give_iterator

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/channels/pushers/buffer.rs:132:21

      <timely::dataflow::stream::StreamCore<S,alloc::vec::Vec<D>> as timely::dataflow::operators::map::Map<S,D>>::map::{{closure}}::{{closure}}::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/map.rs:57:17

      timely::dataflow::operators::generic::handles::InputHandleCore<T,D,P>::for_each

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/handles.rs:97:13

      <timely::dataflow::stream::StreamCore<S,alloc::vec::Vec<D>> as timely::dataflow::operators::map::Map<S,D>>::map::{{closure}}::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/map.rs:55:13

      <timely::dataflow::stream::StreamCore<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::unary::{{closure}}::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/operator.rs:401:17

      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/builder_rc.rs:123:31

      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/builder_rc.rs:162:26

      <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/builder_raw.rs:204:9

   9: timely::progress::subgraph::PerOperatorState<T>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:684:30

  10: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:344:26

      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:310:17

  11: timely::progress::subgraph::PerOperatorState<T>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:684:30

  12: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:344:26

      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:310:17

  13: timely::progress::subgraph::PerOperatorState<T>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:684:30

  14: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:344:26

      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:310:17

  15: timely::worker::Wrapper::step::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/worker.rs:754:57

      core::option::Option<T>::map

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/option.rs:883:29

      timely::worker::Wrapper::step

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/worker.rs:754:26

  16: timely::worker::Worker<A>::step_or_park

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/worker.rs:396:38

  17: mz_dataflow::server::Worker<A,SC,CR>::run

             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0335afc44bffe5852-1/materialize/tests/src/dataflow/src/server.rs:297:13

  18: mz_dataflow::server::serve_boundary::{{closure}}

             at var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0335afc44bffe5852-1/materialize/tests/src/dataflow/src/server.rs:175:9

      timely::execute::execute::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/execute.rs:287:22

      timely_communication::initialize::initialize_from::{{closure}}

             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/communication/src/initialize.rs:316:33

      std::sys_common::backtrace::__rust_begin_short_backtrace

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys_common/backtrace.rs:123:18

  19: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/mod.rs:477:17

      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panic/unwind_safe.rs:271:9

      std::panicking::try::do_call

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:406:40

      std::panicking::try

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:370:19

      std::panic::catch_unwind

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panic.rs:133:14

      std::thread::Builder::spawn_unchecked::{{closure}}

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/mod.rs:476:30

      core::ops::function::FnOnce::call_once{{vtable.shim}}

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/ops/function.rs:227:5

  20: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/alloc/src/boxed.rs:1854:9

      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/alloc/src/boxed.rs:1854:9

      std::sys::unix::thread::Thread::new::thread_start

             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys/unix/thread.rs:108:17

  21: start_thread

  22: clone
```

To be clear, the duplicate records are part of the source database and are not duplicates in the sense of how stream processors worry about. It just so happens for this particular aggregate sum query, I would like to strip out rows with the same xxx_id and pick the latest one by created_at column.

[EDIT] I realized that the `xxx_id` field is a postgres `oid`. Which makes it unique. So it wouldn't make sense to 'dedupe' on that field in queries. The oversight was due to the fact that this was a test database for materialize and I did not thoroughly verify the table schema. Could this be related?

**Comment by jan-wilhelm** [comment]
Same issue here, any update?

For reference, here's my query:
```sql
SELECT DISTINCT ON (c. "lastChangedTime", c.id)
	c.id,
	c. "lastChangedTime",
	m.body
FROM
	"Chat" c
	JOIN "Message" m ON c.id = m. "chatId"
WHERE
	"chatId" = c.id
ORDER BY
	c. "lastChangedTime" DESC,
	c.id DESC
```

**Comment by jan-wilhelm** [comment]
Update: Found an even simpler way to bring materialize to crash:
I'm using a native Postgres cdc connection to a local database that publishes a table called "Message" that includes chat messages.

In materialize, I created a view from the source and run `copy (tail (select max("createdAt") from "Message")) to stdout;`. Now, when I go to the source db and edit the "createdAt" field of the newest message (which should make materialize update its output to the new value), the engine crashes with the following error log, which seems to be *different* from the crash log that @SephVelut and I were running into:
```
2022-10-05T11:33:33.926277Z  INFO mz_dataflow_types::client::controller::storage: timestamp binding feedback persisted in 8.583708ms
2022-10-05T11:33:34.947273Z  INFO mz_dataflow_types::client::controller::storage: timestamp binding feedback persisted in 17.807834ms
2022-10-05T11:33:35.942467Z  INFO mz_dataflow_types::client::controller::storage: timestamp binding feedback persisted in 13.656ms
2022-10-05T11:33:37.924961Z  INFO mz_dataflow_types::client::controller::storage: timestamp binding feedback persisted in 851.083µs
2022-10-05T11:33:38.333994Z ERROR panic: assertion failed: diff > 0
thread: timely:work-0
location: src/dataflow/src/compute/render/reduce.rs:665:13
version: 0.26.4 (8e291a46897cf672db6ac57da309f6cfc71526b5)
backtrace:
   0: materialized::handle_panic
             at var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-0f992ce3704aef3aa-1/materialize/tests/src/materialized/src/bin/materialized/main.rs:1010:21
   1: core::ops::function::Fn::call
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:610:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:500:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys_common/backtrace.rs:139:18
   5: rust_begin_unwind
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:498:5
   6: core::panicking::panic_fmt
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:116:14
   7: core::panicking::panic
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panicking.rs:48:5
   8: mz_dataflow::compute::render::reduce::build_monotonic::{{closure}}
             at var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-0f992ce3704aef3aa-1/materialize/tests/src/dataflow/src/compute/render/reduce.rs:665:13
      <timely::dataflow::stream::StreamCore<S,alloc::vec::Vec<D>> as timely::dataflow::operators::map::Map<S,D>>::map::{{closure}}::{{closure}}::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/map.rs:57:78
      core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &mut F>::call_once
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/ops/function.rs:280:13
   9: core::option::Option<T>::map
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/option.rs:883:29
      <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/iter/adapters/map.rs:103:9
      timely::dataflow::channels::pushers::buffer::Session<T,alloc::vec::Vec<D>,P>::give_iterator
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/channels/pushers/buffer.rs:132:21
      <timely::dataflow::stream::StreamCore<S,alloc::vec::Vec<D>> as timely::dataflow::operators::map::Map<S,D>>::map::{{closure}}::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/map.rs:57:17
      timely::dataflow::operators::generic::handles::InputHandleCore<T,D,P>::for_each
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/handles.rs:97:13
      <timely::dataflow::stream::StreamCore<S,alloc::vec::Vec<D>> as timely::dataflow::operators::map::Map<S,D>>::map::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/map.rs:55:13
      <timely::dataflow::stream::StreamCore<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::unary::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/operator.rs:401:17
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
      timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/builder_rc.rs:162:26
      <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/dataflow/operators/generic/builder_raw.rs:204:9
  10: timely::progress::subgraph::PerOperatorState<T>::schedule
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:684:30
  11: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:310:17
  12: timely::progress::subgraph::PerOperatorState<T>::schedule
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:684:30
  13: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:344:26
      <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/progress/subgraph.rs:310:17
  14: timely::worker::Wrapper::step::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/worker.rs:754:57
      core::option::Option<T>::map
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/option.rs:883:29
      timely::worker::Wrapper::step
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/worker.rs:754:26
  15: timely::worker::Worker<A>::step_or_park
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/worker.rs:396:38
  16: mz_dataflow::server::Worker<A,SC,CR>::run
             at var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-0f992ce3704aef3aa-1/materialize/tests/src/dataflow/src/server.rs:297:13
  17: mz_dataflow::server::serve_boundary::{{closure}}
             at var/lib/buildkite-agent/builds/buildkite-builders-aarch64-585fc7f-i-0f992ce3704aef3aa-1/materialize/tests/src/dataflow/src/server.rs:175:9
      timely::execute::execute::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/timely/src/execute.rs:287:22
      timely_communication::initialize::initialize_from::{{closure}}
             at cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/c4145f0/communication/src/initialize.rs:316:33
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys_common/backtrace.rs:123:18
  18: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/mod.rs:477:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/panic/unwind_safe.rs:271:9
      std::panicking::try::do_call
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:406:40
      std::panicking::try
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panicking.rs:370:19
      std::panic::catch_unwind
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/panic.rs:133:14
      std::thread::Builder::spawn_unchecked::{{closure}}
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/thread/mod.rs:476:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/core/src/ops/function.rs:227:5
  19: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/alloc/src/boxed.rs:1854:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/alloc/src/boxed.rs:1854:9
      std::sys::unix::thread::Thread::new::thread_start
             at rustc/9d1b2106e23b1abd32fce1f17267604a5102f57a/library/std/src/sys/unix/thread.rs:108:17
  20: start_thread
  21: <unknown>
```

Note: This does not happen when editing unrelated fields in the Message table that aren't needed to produce the result set.

**Comment by frankmcsherry** [comment]
Fwiw, this appears to be an instance of some part of Materialize being instructed that the input will not contain retractions (the errors are in `render_top1_monotonic` and `reduce::build_monotonic`), for an input that nonetheless supports retractions. My best guess is that this code retains a defect we fixed where anything that advertises itself as `ENVELOPE NONE` is treated as "append-only", and new sources were added without updating this logic.

**Comment by jan-wilhelm** [comment]
> Fwiw, this appears to be an instance of some part of Materialize being instructed that the input will not contain retractions (the errors are in `render_top1_monotonic` and `reduce::build_monotonic`), for an input that nonetheless supports retractions. My best guess is that this code retains a defect we fixed where anything that advertises itself as `ENVELOPE NONE` is treated as "append-only", and new sources were added without updating this logic.

That makes sense to me after reading the docs - is there anything I can do myself (e..g when creating the Postgres source in materialize) to force Materialize to use an upsert envelope?

**Comment by uce** [comment]
@jan-wilhelm Unfortunately, there is nothing you can do as a user. We'll backport the fix next week and the issue should be resolved with the next LTS release of the binary (I can't commit on an actual release date though).

**Comment by jan-wilhelm** [comment]
Great, thank you! Looking forward to the release.

**Comment by benesch** [comment]
Released this afternoon as v0.26.5. Thanks again for the report, and sorry about the bug in the first place! Going to preemptively close this out, but give us a holler if something still seems broken.
